File: user_management_service/saga_coordinator.py
import requests
import logging

logger = logging.getLogger('saga_coordinator')

class SagaCoordinator:
    def __init__(self):
        self.steps = []
        self.compensation_steps = []

    def add_step(self, step, compensation_step):
        self.steps.append(step)
        self.compensation_steps.insert(0, compensation_step)

    def execute(self):
        for step in self.steps:
            try:
                step()
            except Exception as e:
                logger.error(f"Step failed: {e}")
                self.rollback()
                raise e

    def rollback(self):
        for compensation_step in self.compensation_steps:
            try:
                compensation_step()
            except Exception as e:
                logger.error(f"Compensation step failed: {e}")

def create_user_step(user_data):
    def step():
        response = requests.post('http://gateway:5000/users/user/register', json=user_data)
        response.raise_for_status()
        user_data['_id'] = response.json()['user']['_id']
    return step

def create_game_session_step(game_data):
    def step():
        response = requests.post('http://gateway:5000/game/start-game', json=game_data)
        response.raise_for_status()
        game_data['_id'] = response.json()['game']['_id']
    return step

def delete_user_step(user_data):
    def step():
        response = requests.delete(f'http://gateway:5000/users/{user_data["_id"]}')
        response.raise_for_status()
    return step

def delete_game_session_step(game_data):
    def step():
        response = requests.delete(f'http://gateway:5000/game/{game_data["_id"]}')
        response.raise_for_status()
    return step

File: gateway/saga_coordinator.js
const axios = require('axios');

class SagaCoordinator {
    constructor() {
        this.steps = [];
        this.compensation_steps = [];
    }

    addStep(step, compensationStep) {
        this.steps.push(step);
        this.compensation_steps.unshift(compensationStep);
    }

    async execute() {
        for (const step of this.steps) {
            try {
                await step();
            } catch (e) {
                console.error(`Step failed: ${e}`);
                await this.rollback();
                throw e;
            }
        }
    }

    async rollback() {
        for (const compensationStep of this.compensation_steps) {
            try {
                await compensationStep();
            } catch (e) {
                console.error(`Compensation step failed: ${e}`);
            }
        }
    }
}

const createUserStep = (userData) => async () => {
    const response = await axios.post('http://gateway:5000/users/user/register', userData);
    userData._id = response.data.user._id;
};

const deleteUserStep = (userData) => async () => {
    await axios.delete(`http://gateway:5000/users/${userData._id}`);
};

const createGameSessionStep = (gameData) => async () => {
    const response = await axios.post('http://gateway:5000/game/start-game', gameData);
    gameData._id = response.data.game._id;
};

const deleteGameSessionStep = (gameData) => async () => {
    await axios.delete(`http://gateway:5000/game/${gameData._id}`);
};

module.exports = {
    SagaCoordinator,
    createUserStep,
    deleteUserStep,
    createGameSessionStep,
    deleteGameSessionStep
};